File: swockets.NET/swockets.NET/Swockets.cs
using System;
using System.Net.Sockets;
using System.Net;
using System.IO;
using System.Text;
using System.Threading;
using Newtonsoft.Json.Linq;
using System.Collections.Generic;

namespace swockets
{
	public class SwocketError : Exception
	{
		public SwocketError()
		{

		}

		public SwocketError(string value)
		{
		}
	}

	public class SwocketClientSocket
	{
		public Socket sock;
		public string address;

		public SwocketClientSocket(Socket sock, string address)
		{
			this.sock = sock;
			this.address = address;
		}
	}

	public class SwocketHandler
	{
		/*
		 * 
		 */
		public bool handshake(Socket sock)
		{
			return true;
		}

		/*
		 * 
		 */
		public void recv(JObject recvObj, SwocketClientSocket sock = null)
		{
			Console.WriteLine(recvObj);
		}

		/*
		 *
		 */
		public void connect(SwocketClientSocket sock = null)
		{
			Console.WriteLine("connect");
		}

		/*
		 * 
		 */
		public void disconnect(SwocketClientSocket sock = null)
		{
			Console.WriteLine("disconnect");
		}

		/*
		 * 
		 */
		public void handshake_unsuccessful()
		{
			Console.WriteLine("handshake unsuccessful");
		}
	}

	public enum SwocketMode
	{
		ISSERVER, ISCLIENT
	}

	public class Swockets
	{
		private SwocketHandler handler;
		private SwocketMode mode;
		private Socket sock;
		private Thread serverConnectionThread;
		private bool RUNNING = true;
		private List<SwocketClientSocket> clients;

		public Swockets(SwocketMode mode, SwocketHandler handler, string host, int port = 6666, int backlog = 1)
		{
			this.mode = mode;
			this.handler = handler;


			//initialize as Swockets Server
			if (this.mode == SwocketMode.ISSERVER)
			{
				IPEndPoint localEndPoint = new IPEndPoint(IPAddress.Parse("127.0.0.1"), port);
				sock = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);

				sock.Bind(localEndPoint);
				sock.Listen(backlog);

				clients = new List<SwocketClientSocket>();

				serverConnectionThread = new Thread(server_connection_thread);
				serverConnectionThread.Start();
			}
			//initialize as Swockets Client
			else
			{
				IPEndPoint ipe = new IPEndPoint(IPAddress.Parse(host), port);
				sock = new Socket(ipe.AddressFamily, SocketType.Stream, ProtocolType.Tcp);

				sock.Connect(ipe);

				if (sock.Connected)
				{
					client_negotiate();
				}
			}
		}

		/*
		 * 
		 */
		public void stop()
		{
			RUNNING = false;

			if (mode == SwocketMode.ISSERVER)
			{
				foreach (SwocketClientSocket client in clients)
				{
					client.sock.Dispose();
				}
			}
			sock.Dispose();
		}

		/*
		 * 
		 */
		private void server_connection_thread()
		{
			while (RUNNING)
			{
				Socket handler = sock.Accept();
				server_negotiate(handler, ((IPEndPoint)handler.RemoteEndPoint).Address.ToString());
			}
		}

		/*
		 * 
		 */
		private void server_negotiate(Socket clientsocket, string address)
		{
			if (handler.handshake(clientsocket))
			{
				SwocketClientSocket client = new SwocketClientSocket(clientsocket, address);
				handler.connect(client);
				clients.Add(client);

				Thread t = new Thread(() => receive_thread(clientsocket, client));
				t.Start();
			}
			else
			{
				clientsocket.Dispose();
				handler.handshake_unsuccessful();
			}
		}

		/*
		 * 
		 */
		private void client_negotiate()
		{
			if (handler.handshake(sock))
			{
				handler.connect(new SwocketClientSocket(sock, ""));
				Thread t = new Thread(() => receive_thread(sock));
				t.Start();
			}
			else
			{
				sock.Dispose();
				handler.handshake_unsuccessful();
			}

		}

		/*
		 * 
		 */
		private JObject receive_one_message(Socket sock, SwocketClientSocket clsock = null)
		{
			string recvdMsg = "";

			while (true)
			{
				try
				{
					Byte[] bytesReceived = new Byte[1024];

					int bytes = sock.Receive(bytesReceived);
					recvdMsg += Encoding.ASCII.GetString(bytesReceived, 0, bytes);

					JObject recvdObj = JObject.Parse(recvdMsg);
					recvdMsg = "";

					return recvdObj;
				}
				catch (Newtonsoft.Json.JsonReaderException)
				{
					//continue receiving
				}
				catch (Exception e)
				{
					Console.Error.WriteLine("Error while receiving message: " + e);
					if (mode == SwocketMode.ISSERVER)
					{
						clients.Remove(clsock);
						sock.Dispose();
						handler.disconnect(clsock);
					}
					else
					{
						this.sock.Dispose();
						handler.disconnect();
					}
					return null;
				}
			}

		}

		/*
		 * 
		 */
		private void receive_thread(Socket sock, SwocketClientSocket clsock = null)
		{
			while (RUNNING)
			{
				JObject recvdObj = receive_one_message(sock, clsock);

				if (recvdObj == null)
				{
					return;
				}
				handler.recv(recvdObj, clsock);
			}
		}

		/*
		 * 
		 */
		public JObject receive(Socket sock)
		{
			return receive_one_message(sock);
		}

		/*
		 * 
		 */
		public void send(JObject msg, Socket sock = null, SwocketClientSocket clsock = null)
		{
			if (mode == SwocketMode.ISCLIENT)
			{
				sock = this.sock;
			}
			try
			{
				string sendStr = msg.ToString();

				if (sendStr != "")
				{
					sock.Send(Encoding.ASCII.GetBytes(msg.ToString() + new string(' ', (1024 - (System.Text.ASCIIEncoding.ASCII.GetByteCount(msg.ToString())) % 1024))));
				}
			}
			catch (Exception e)
			{
				Console.Error.WriteLine("Error while receiving message: " + e);
				if (mode == SwocketMode.ISSERVER)
				{
					clients.Remove(clsock);
					sock.Dispose();
					handler.disconnect(clsock);
				}
				else
				{
					this.sock.Dispose();
					handler.disconnect();
				}
			}
		}
	}
}

File: swockets.py/swockets.py
#!/usr/bin/env python

"""
swockets by Daniel Swoboda (@snoato, swobo.space)
"""

import socket
import threading
import json
import select
import sys

#Exception for swockets
class SwocketError(IOError):
	def __init__(self, value):
		self.value = value

	def __str__(self):
		return self.value

#Class to store necessary client data for swockets in server mode
class SwocketClientSocket():
	pass

#Basic Handler as it can be used for swockets
class SwocketHandler(object):
	def __init__(self):
		self.sock = None
		self.connected = None

	#if in server mode, sock = SwocketClientSocket of connecting socket
	def handshake(self, sock):
		return True

	#if in server mode, sock = SwocketClientSocket of sending socket
	def recv(self, recvObj, sock = None):
		print recvObj

	#if in server mode, sock = SwocketClientSocket of disconnected socket
	def disconnect(self, sock = None):
		print "disconnect"

	def connect(self, sock):
		print "connect"

	def handshake_unsuccessful(self):
		print "handshake unsuccessful"


#Swockets, the Sw(obo - S)ockets
class swockets:
	ISSERVER = 1
	ISCLIENT = 2
	RUNNING = True

	def __init__(self, mode, handler, host=None, port=6666, backlog=1):
		self.handler = handler
		self.mode = mode
		self.handler.sock = self

		if mode == swockets.ISSERVER:
			self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
			self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1) 
			self.sock.bind(('', port))
			self.sock.listen(backlog)
			self.clients = []

			t=threading.Thread(target=self.server_connection_thread, args=())
			t.daemon = True  
			t.start()

		elif mode == swockets.ISCLIENT:
			self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
			self.sock.connect((host, port))

			self.client_negotiate()
		else:
			raise ValueError("You must provide a correct mode")

	def __del__(self):
		self.stop()

	#closes all connections and stops all running loops
	def stop(self):
		self.RUNNING = False
		if self.mode == swockets.ISSERVER:
			for client in self.clients:
				client.sock.close()
				
		self.sock.close()

	#thread that accepts new connections for server
	def server_connection_thread(self):
		while self.RUNNING:
			(clientsocket, address) = self.sock.accept()
			self.server_negotiate(clientsocket, address)

	#server negotiating function
	def server_negotiate(self, clientsocket, address):
		if self.handler.handshake(clientsocket):
			client = SwocketClientSocket()
			client.sock = clientsocket
			client.address = address
			self.clients.append(client)
			self.handler.connect(clientsocket)

			t=threading.Thread(target=self.receive_thread, args=(client, client.sock))
			t.daemon = True  
			t.start()
		else:
			clientsocket.close()
			self.handler.handshake_unsuccessful()

	#client negotiating function
	def client_negotiate(self):
		if self.handler.handshake(self.sock):
			self.handler.connect(self.sock)
			t=threading.Thread(target=self.receive_thread, args=(self.sock, self.sock))
			t.daemon = True  
			t.start()
		else:
			self.sock.close()
			self.handler.handshake_unsuccessful()

	#receives one json formatted message between two swockets
	def receive_one_message(self, sock, ssock):
		recvdStr = ""
		recvdMsg = ""

		while True:
			try:
				recvdStr = ssock.recv(1024)

				if recvdStr == "":
					if self.mode == swockets.ISSERVER:
						self.handler.disconnect(sock)
						ssock.close()
						self.clients.remove(sock)
						return None
					else: 
						self.handler.disconnect()
						self.sock.close()
						return None
				else:
					recvdMsg+=recvdStr

				recvdObj = json.loads(recvdMsg)
				recvdMsg = ""

				return recvdObj
			except ValueError:
				pass
			except socket.error:
				if self.mode == swockets.ISSERVER:
					ssock.close()
					try:
						self.handler.disconnect(sock)
						self.clients.remove(sock)
					except:
						pass
					return None
				else:
					self.handler.disconnect()
					self.sock.close()
					return None

	#thread that runs in background waiting for messages and forwarding
	#them to the handler for further processing by user
	def receive_thread(self, sock = None, ssock = None):
		if self.mode == swockets.ISCLIENT:
			sock = self.sock
			ssock = self.sock

		while self.RUNNING:
			recvdObj = self.receive_one_message(sock, ssock)

			if recvdObj == None:
				return
			self.handler.recv(recvdObj, sock)

		sock.sock.close()

	#user callable receive function, provide sock if in server mode
	def receive(self, sock = None, ssock = None):
		if self.mode == swockets.ISCLIENT:
			ssock = self.sock

		return self.receive_one_message(sock, ssock)

	#user callable send function, provide sock if in server mode, 
	#msg has to be json object
	#sock and ssock have to be given in server mode
	def send(self, msg, sock = None, ssock = None):
		try:	
			if self.mode == swockets.ISCLIENT:
				ssock = self.sock

			sendStr = json.dumps(msg)

			if sendStr != "":
				sendStr += ' ' * (1024-(len(sendStr.encode("utf8")) % 1024))
				ssock.sendall(sendStr)
		except socket.error:
			if self.mode == swockets.ISCLIENT:
				self.handler.disconnect()
			else:
				self.handler.disconnect(sock)
		except IndexError:
			if sock not in self.clients:
				SwocketError("Client not available")

